File: jython/pythonLib/pmask/oscillator.py
# oscillator.py

# PMask, a Python implementation of CMask
# Copyright (C) 2000 by Maurizio Umberto Puxeddu

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

from math import pow, fabs, fmod, sin, cos, pi
from generator import *

class Sine(Generator):
    def __init__(self, T = 1.0, ph = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        ph = self.ph + evaluationTime / t
        return (1.0 + sin(2.0 * pi * ph)) / 2.0

class Cosine(Generator):
    def __init__(self, T = 1.0, ph = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        ph = self.ph + evaluationTime / t
        return (1.0 + cos(2.0 * pi * ph)) / 2.0

class SawUp(Generator):
    def __init__(self, T = 1.0, ph = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        evaluationTime += t * self.ph
        evaluationTime = fmod(evaluationTime, t)
        if evaluationTime < 0:
            evaluationTime += t
        return evaluationTime / t

class SawDown(Generator):
    def __init__(self, T = 1.0, ph = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        evaluationTime += t * self.ph
        evaluationTime = fmod(evaluationTime, t)
        if evaluationTime < 0:
            evaluationTime += t
        return 1.0 - evaluationTime / t

class PowerUp(Generator):
    def __init__(self, T = 1.0, ph = 0.0, exponent = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph
        self.exponent = pow(2.0, exponent)

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        evaluationTime += t * self.ph
        evaluationTime = fmod(evaluationTime, t)
        if evaluationTime < 0:
            evaluationTime += t

        return pow(evaluationTime / t, self.exponent)
        
class PowerDown(Generator):
    def __init__(self, T = 1.0, ph = 0.0, exponent = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph
        self.exponent = pow(2.0, exponent)

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        evaluationTime += t * self.ph
        evaluationTime = fmod(evaluationTime, t)
        if evaluationTime < 0:
            evaluationTime += t
        return pow(1.0 - evaluationTime / t, self.exponent)

class Square(Generator):
    def __init__(self, T = 1.0, ph = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        evaluationTime += t * self.ph
        evaluationTime = fmod(evaluationTime, t)
        if evaluationTime < 0:
            evaluationTime += t
        if evaluationTime < t/2:
            return 1.0
        else:
            return 0

class Triangle(Generator):
    def __init__(self, T = 1.0, ph = 0.0):
        Generator.__init__(self)
        self.T = T
        self.ph = ph

    def valueAt(self, evaluationTime):
        t = evaluateAt(self.T, evaluationTime)
        evaluationTime += t * self.ph
        t_2 = t / 2
        evaluationTime = fmod(evaluationTime, t)
        if evaluationTime < 0:
            evaluationTime += t
        if evaluationTime < t_2:
            return evaluationTime / t_2
        else:
            return 1.0 - (evaluationTime - t_2)/t_2

if __name__ == '__main__':
    f1 = Sine(1.0)
    f2 = Cosine(1.0)
    f3 = SawUp(1.0)
    f4 = SawDown(1.0)
    f5 = PowerUp(1.0, 0, 1.0)
    f6 = PowerUp(1.0, 0, 2.0)
    f7 = PowerUp(1.0, 0, -1.0)
    f8 = PowerDown(1.0, 0, 1.0)
    f9 = PowerDown(1.0, 0, 2.0)
    f10 = PowerDown(1.0, 0, -1.0)
    f11 = Square(1.0)
    f12 = Triangle(1.0)
    f13 = Sine(1.0, 0.3)
    f14 = Cosine(1.0, 0.3)
    f15 = SawUp(1.0, 0.3)
    f16 = PowerUp(1.0, 0.3, 2.0)
    f17 = PowerDown(1.0, 0.3, 2.0)
    f18 = Square(1.0, 0.3)
    f19 = Triangle(1.0, 0.3)
    
    t = -2.0
    step = 0.01
    for i in range(401):
        print t, f1.valueAt(t), f2.valueAt(t), f3.valueAt(t),\
              f4.valueAt(t), f5.valueAt(t), f6.valueAt(t),\
              f7.valueAt(t), f8.valueAt(t), f9.valueAt(t),\
              f10.valueAt(t), f11.valueAt(t), f12.valueAt(t),\
              f13.valueAt(t), f14.valueAt(t), f15.valueAt(t),\
              f16.valueAt(t), f17.valueAt(t), f18.valueAt(t), f19.valueAt(t)
        t += step

File: blue-core/src/blue/soundObject/jmask/Oscillator.java
/*
 * blue - object composition environment for csound
 * Copyright (c) 2007 Steven Yi (stevenyi@gmail.com)
 *
 * Based on CMask by Andre Bartetzki
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published
 * by  the Free Software Foundation; either version 2 of the License or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.LIB.  If not, write to
 * the Free Software Foundation Inc., 59 Temple Place - Suite 330,
 * Boston, MA  02111-1307 USA
 */
package blue.soundObject.jmask;

import blue.utility.MathUtils;
import blue.utility.XMLUtilities;
import electric.xml.Element;
import electric.xml.Elements;
import java.io.Serializable;
import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.ToStringBuilder;

public class Oscillator implements Generator, Serializable, Maskable,
        Quantizable, Accumulatable {

    public static final int SINE = 0;

    public static final int COSINE = 1;

    public static final int SAW_UP = 2;

    public static final int SAW_DOWN = 3;

    public static final int SQUARE = 4;

    public static final int TRIANGLE = 5;

    public static final int POW_UP = 6;

    public static final int POW_DOWN = 7;

    private static final double PI2 = Math.PI * 2;

    public static final String[] FUNCTIONS = { "Sine", "Cosine",
            "Saw (Increasing)", "Saw (Decreasing)", "Square", "Triangle",
            "Power Function (Increasing)", "Power Function (Decreasing)" };

    private int oscillatorType = SINE;

    double phaseInit = 0.0;

    private double frequency = 1.0;

    private Table freqTable = new Table();

    private boolean freqTableEnabled = false;

    double exponent = 1.0;
    
    public Oscillator() {
        freqTable.setMin(.001, false);
        freqTable.setMax(10, false);
        
        TablePoint point1 = (TablePoint) freqTable.points.get(0);
        TablePoint point2 = (TablePoint) freqTable.points.get(1);
        
        point1.setValue(1);
        point2.setValue(1);
    }

    public static Generator loadFromXML(Element data) {
        Oscillator retVal = new Oscillator();

        Elements nodes = data.getElements();

        while (nodes.hasMoreElements()) {
            Element node = nodes.next();
            String nodeName = node.getName();
            switch (nodeName) {
                case "oscillatorType":
                    retVal.oscillatorType = Integer.parseInt(node.getTextString());
                    break;
                case "phaseInit":
                    retVal.phaseInit = Double.parseDouble(node.getTextString());
                    break;
                case "frequency":
                    retVal.setFrequency(Double.parseDouble(node.getTextString()));
                    break;
                case "freqTableEnabled":
                    retVal.setFreqTableEnabled(Boolean
                            .valueOf(node.getTextString()).booleanValue());
                    break;
                case "table":
                    retVal.setFreqTable(Table.loadFromXML(node));
                    break;
                case "exponent":
                    retVal.exponent = Double.parseDouble(node.getTextString());
                    break;
            }

        }

        return retVal;
    }

    public Element saveAsXML() {
        Element retVal = new Element("generator");
        retVal.setAttribute("type", getClass().getName());

        retVal.addElement(XMLUtilities.writeInt("oscillatorType",
                oscillatorType));
        retVal.addElement(XMLUtilities.writeDouble("phaseInit", phaseInit));
        retVal
                .addElement(XMLUtilities.writeDouble("frequency",
                        getFrequency()));
        retVal.addElement(XMLUtilities.writeBoolean("freqTableEnabled",
                isFreqTableEnabled()));
        retVal.addElement(getFreqTable().saveAsXML());
        retVal.addElement(XMLUtilities.writeDouble("exponent", exponent));

        return retVal;
    }

//    public JComponent getEditor() {
//        return new OscillatorEditor(this);
//    }

    public double getExponent() {
        return exponent;
    }

    public void setExponent(double exponent) {
        this.exponent = exponent;
    }

    public int getOscillatorType() {
        return oscillatorType;
    }

    public void setOscillatorType(int oscillatorType) {
        this.oscillatorType = oscillatorType;
    }

    public void initialize(double duration) {
//        freqTable.setInitialPhase(phaseInit);
    }

    public double getValue(double time, java.util.Random rnd) {
        double retVal = 0.0;

        double phase = getPhase(time);

        switch (oscillatorType) {
            case SINE:
                retVal = sin(phase);
                break;
            case COSINE:
                retVal = cos(phase);
                break;
            case SAW_UP:
                retVal = sawup(phase);
                break;
            case SAW_DOWN:
                retVal = sawdown(phase);
                break;
            case SQUARE:
                retVal = square(phase);
                break;
            case TRIANGLE:
                retVal = triangle(phase);
                break;
            case POW_UP:
                retVal = powerup(phase);
                break;
            case POW_DOWN:
                retVal = powerdown(phase);
                break;
        }

        return retVal;
    }

    private double getPhase(double time) {
        double retVal = 0.0;
        
        if (!freqTableEnabled) {
            retVal = phaseInit + (time * frequency);
        } else {
            retVal = phaseInit + freqTable.getphs(time);
        }

        return retVal;
    }

    public double getPhaseInit() {
        return phaseInit;
    }

    public void setPhaseInit(double phaseInit) {
        this.phaseInit = phaseInit;
    }

    private double sin(double phpt) {
        return Math.sin(PI2 * phpt) * 0.5 + 0.5;
    }

    private double cos(double phpt) {
        return Math.cos(PI2 * phpt) * 0.5 + 0.5;
    }

    private double sawup(double phpt) {
        System.err.println("Phase: " + phpt + " Value: "
                + MathUtils.remainder(phpt, 1.0));
        return Math.abs(MathUtils.remainder(phpt, 1.0));
    }

    private double sawdown(double phpt) {
        return 1.0 - Math.abs(MathUtils.remainder(phpt, 1.0));
    }

    private double square(double phpt) {
        double x;
        x = Math.abs(MathUtils.remainder(phpt, 1.0));
        return (x < 0.5 ? 1.0 : 0.0);
    }

    private double triangle(double phpt) {
        double x;
        x = Math.abs(MathUtils.remainder(phpt, 1.0));
        return (x < 0.5 ? (2.0 * x) : (2.0 * (1.0 - x)));
    }

    private double powerup(double phpt) {
        return Math.pow(Math.abs(MathUtils.remainder(phpt, 1.0)), Math.pow(2.0,
                exponent));
    }

    private double powerdown(double phpt) {
        return Math.pow(1.0 - Math.abs(MathUtils.remainder(phpt, 1.0)), Math
                .pow(2.0, exponent));
    }

    @Override
    public boolean equals(Object obj) {
        return EqualsBuilder.reflectionEquals(this, obj);
    }

    @Override
    public String toString() {
        return ToStringBuilder.reflectionToString(this);
    }

    public double getFrequency() {
        return frequency;
    }

    public void setFrequency(double frequency) {
        this.frequency = frequency;
    }

    public Table getFreqTable() {
        return freqTable;
    }

    public void setFreqTable(Table freqTable) {
        this.freqTable = freqTable;
    }

    public boolean isFreqTableEnabled() {
        return freqTableEnabled;
    }

    public void setFreqTableEnabled(boolean freqTableEnabled) {
        this.freqTableEnabled = freqTableEnabled;
    }
}
